File: DCP/DCP/Day 9/Day9.swift


func largestSum(numbers: [Int]) -> Int {
    guard !numbers.isEmpty else { return 0 }

    guard numbers.count != 1 else { return numbers.first! }

    var prevMax = numbers[0]
    var currMax = max(numbers[0], numbers[1])
    for i in 2..<numbers.count {
        let temp = currMax
        currMax = max(currMax, prevMax + numbers[i])
        prevMax = temp
    }
    return currMax
}

import XCTest

final class Day9: XCTestCase {
    func test() {
        XCTAssertEqual(largestSum(numbers: [2, 4, 6, 2, 5]), 13)
        XCTAssertEqual(largestSum(numbers: [5, 1, 1, 5]), 10)
    }
}

File: DCP/DCP/Day 9/Day9.kt
fun main() {
    println(largestSum(listOf(2, 4, 6, 2, 5)) == 13)
    println(largestSum(listOf(5, 1, 1, 5)) == 10)
}

fun largestSum(numbers: List<Int>): Int {
    if (numbers.isEmpty()) return 0
    if (numbers.size == 1) return numbers[0]
    
    var prevMax = numbers[0]
    var currMax = maxOf(numbers[0], numbers[1])
    for (i in 2 until numbers.size) {
        val temp = currMax
        currMax = maxOf(currMax, prevMax + numbers[i])
        prevMax = temp
    }
    
    return currMax
}